openlibrary/templates/site/body.html

```diff
@@ -29,12 +29,14 @@
       $for flash in get_flash_messages():
         <div class="$flash.type"><span>$flash.message</span></div>
       </div>
-      $# Paste next line into announcement variable to show the blue banner
-      $# <strong>Title:</strong> Description &nbsp; <a href="#" data-ol-link-track="AnnouncementsBar|BlogWidget" class="btn primary">Learn More</a>
-      $ announcement = None
-      $if announcement:
-        <div class="page-banner page-banner-body">
-         $:announcement
-        </div>
+      $# Announcement banner will only be rendered if announcement and storage_key variables are set.
+      $# Be sure to escape any single quotes inside of the announcement HTML string.
+      $ announcement = ''
+      $# Using the current date in YYYY-MM-DD format for storage_key will prevent the use of duplicate keys.
+      $ storage_key = ''
+      $ hide_banner = ctx.user and ctx.user.preferences().get('hidden-banner') == storage_key
+      $if announcement and storage_key:
+        $if not hide_banner:
+          $:render_template('site/banner', announcement, storage_key)
```

Comment:
```suggestion
      $if announcement and storage_key:
        $if ctx.user:
          $if ctx.user.preferences().get('hidden-banner') != storage_key:
            $:render_template('site/banner', announcement, storage_key)
        $else:
          $:render_template('site/banner', announcement, storage_key)
```
The `hide_banner` assignment is fetching patron's preferences even if there is no banner.  Also, `ctx.user` shouldn't exist when the patron is logged out, so the `preferences()` call should throw an error (I'm very unclear as to why a banner was present when testing while logged out...).

Reply:
I believe the `preferences()` call will not occur if the first boolean check (`ctx.user`) is already returning `False`